plio/io/io_bae.py

```diff
@@ -126,6 +125,54 @@ def read_ipf_list(input_data_list):
 
     return df
 
+def save_ipf(df, output_path):
+    """
+    Write a socet gpf file from a gpf-defined pandas dataframe
+
+    Parameters
+    ----------
+    df          : pd.DataFrame
+                  Pandas DataFrame
+
+    output_file : str
+                  path to the output data file
+
+    Returns
+    -------
+    int         : success value
+                  0 = success, 1 = errors
+    """
+
+    for ipf_file, ipf_df in df.groupby('ipf_file'):
+
+        output_file = os.path.join(output_path, ipf_file + '.ipf')
+
+        # Check that file can be opened
+        try:
+            outIPF = open(output_file, 'w', newline='\r\n')
+        except:
+            print('Unable to open output ipf file: {0}'.format(output_file))
+            return 1
+
+        #grab number of rows in pandas dataframe ipf group
+        numpts = len(ipf_df)
+
+        #Output ipf header
+        outIPF.write('IMAGE POINT FILE\n')
+        outIPF.write('{0}\n'.format(numpts))
+        outIPF.write('pt_id,val,fid_val,no_obs,l.,s.,sig_l,sig_s,res_l,res_s,fid_x,fid_y\n')
+
+        for index, row in ipf_df.iterrows():
+            #Output coordinates to ipf file
+            outIPF.write('{0} {1} {2} {3}\n'.format(row['pt_id'], row['val'], row['fid_val'], row['no_obs']))
+            outIPF.write('{:0.6f}  {:0.6f}\n'.format(row['l.'], row['s.']))
+            outIPF.write('{:0.6f}  {:0.6f}\n'.format(row['sig_l'], row['sig_s']))
+            outIPF.write('{:0.6f}  {:0.6f}\n'.format(row['res_l'], row['res_s']))
+            outIPF.write('{:0.6f}  {:0.6f}\n\n'.format(row['fid_x'], row['fid_y']))
```

Comment:
@dpmayerUSGS @thareUSGS Checkout the sweet formatting magic! :wink:

Reply:
In practical terms, 6 decimal places is plenty for IPF coordinates, but I vote to make as faithful a translation as possible. Can we adjust the formatting so that we preserve whatever the precision is in each item from the data frame when we go to write the new IPF file?

Reply:
I suspect, but @acpaquette chime in, that will require also tracking the precision that we get in from socet inside of some other data structure. While nice, this might not be practical?

Reply:
Do we really need to track that separately? 
Take a look at `save_gpf()`. It just copies numbers from the data frame "as-is" into the new GPF file.

Reply:
in that func we can use `genfromtxt` which gets us that tracking for free.  As far as I know, this is not possible with the ipf due to the strange formating of each entry.

Reply:
There may actually be the opposite issue for certain fields. val, fid_val, no_obs should all be downcast to unsigned integers. Perhaps something like:
`df.val = pf.to_numeric(df['val'], downcast = 'unsigned')`

Reply:
I was mainly trying to return exactly what the data looks like in the current ipfs. If you dont do the formatting there are erroneous floating point values that return like 10.444040000000001.

Reply:
Ok, 6 digits works for me then.